cpp/src/arrow/compute/kernels/take.cc

```diff
@@ -0,0 +1,351 @@
+  if (type_id != Type::NA) {
+    const auto& fw_type = static_cast<const FixedWidthType&>(*out->type);
+    int bit_width = fw_type.bit_width();
+    int64_t buffer_size = 0;
+    if (bit_width == 1) {
+      buffer_size = BitUtil::BytesForBits(length);
+    } else if (bit_width % 8 == 0) {
+      buffer_size = length * fw_type.bit_width() / 8;
+    } else {
+      DCHECK(false);
+    }
+    RETURN_NOT_OK(ctx->Allocate(buffer_size, &out_data));
+    memset(out_data->mutable_data(), 0, buffer_size);
+  }
+
+  out->buffers.push_back(validity_bitmap);
+  out->buffers.push_back(out_data);
+
+  return Status::OK();
+}
+
+template <typename IndexType>
+class TakeKernel : public UnaryKernel {};
+
+template <>
+class TakeKernel<Int32Type> : public UnaryKernel {
+ public:
+  TakeKernel(const TakeOptions& options, const TakeFunction& func,
+             const std::shared_ptr<DataType>& index_type)
+      : options_(options),
+        func_(func),
+        index_type_(index_type) {}
+
+  Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
+    DCHECK_EQ(Datum::COLLECTION, input.kind());
+
+    const std::vector<Datum>& input_datums = input.collection();
+    auto input_array = input_datums[0].array();
+    auto indies_array = input_datums[1].array();
+
+    ArrayData* result;
+    if (out->kind() == Datum::NONE) {
+      out->value = ArrayData::Make(input_array->type, indies_array->length);
+    }
+    result = out->array().get();
+    RETURN_NOT_OK(AllocateIfNotPreallocated(ctx, indies_array->length, result));
+
+    func_(ctx, options_, *input_array, *indies_array, result);
+
+    RETURN_IF_ERROR(ctx);
+    return Status::OK();
+  }
+
+ private:
+  TakeOptions options_;
+  TakeFunction func_;
+  std::shared_ptr<DataType> index_type_;
+};
+
+#define TAKE_CASE(ValueType, IndexType)                                                 \
+  case IndexType::type_id:                                                              \
+    func = [](FunctionContext* ctx, const TakeOptions& options, const ArrayData& input, \
+              const ArrayData& indices, ArrayData* out) {                               \
+      TakeFunctor<ValueType, IndexType> func;                                           \
+      func(ctx, options, input, indices, out);                                          \
+    };                                                                                  \
+    break;
+
+#define PRIMITIVE_CASES(FN, VALUE_TYPE) \
+  FN(VALUE_TYPE, Int8Type);           \
+  FN(VALUE_TYPE, UInt8Type);          \
+  FN(VALUE_TYPE, Int16Type);          \
+  FN(VALUE_TYPE, UInt16Type);         \
+  FN(VALUE_TYPE, Int32Type);          \
+  FN(VALUE_TYPE, UInt32Type);         \
+  FN(VALUE_TYPE, Int64Type);          \
+  FN(VALUE_TYPE, UInt64Type);
+
+#define GET_TAKE_FUNCTION(CASE_GENERATOR, ValueType)                           \
+  static std::unique_ptr<UnaryKernel> Get##ValueType##TakeFunc(                \
```

Comment:
I'd make this function return Status and pass the unique_ptr by output parameter (as the parent callsite is also doing).